cmd/transcriber/apis/azure/speech_recognizer.go

```diff
@@ -0,0 +1,287 @@
+	})
+	speechRecognizer.Canceled(func(event speech.SpeechRecognitionCanceledEventArgs) {
+		defer event.Close()
+		slog.Info("transcription canceled", slog.String("details", event.ErrorDetails))
+	})
+	speechRecognizer.Recognizing(func(event speech.SpeechRecognitionEventArgs) {
+		defer event.Close()
+		slog.Info("recognizing", slog.Any("result", event.Result))
+	})
+
+	return speechRecognizer, audioConfig, audioStream, nil
+}
+
+func NewSpeechRecognizer(cfg SpeechRecognizerConfig) (*SpeechRecognizer, error) {
+	if err := cfg.IsValid(); err != nil {
+		return nil, fmt.Errorf("failed to validate config: %w", err)
+	}
+
+	speechConfig, err := speech.NewSpeechConfigFromSubscription(cfg.SpeechKey, cfg.SpeechRegion)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create speech config: %w", err)
+	}
+	if err := speechConfig.SetProperty(common.SpeechLogFilename, filepath.Join(cfg.DataDir, "azure.log")); err != nil {
+		return nil, fmt.Errorf("failed to set log property: %w", err)
+	}
+
+	speechRecognizer, audioConfig, audioStream, err := initSpeechRecognizer(speechConfig)
+	if err != nil {
+		return nil, err
+	}
+
+	sr := &SpeechRecognizer{
+		cfg:              cfg,
+		speechConfig:     speechConfig,
+		speechRecognizer: speechRecognizer,
+		audioConfig:      audioConfig,
+		audioStream:      audioStream,
+	}
+
+	return sr, nil
+}
+
+func (s *SpeechRecognizer) TranscribeAsync(samplesCh <-chan []float32) (<-chan transcribe.Segment, error) {
+	segmentsCh := make(chan transcribe.Segment, 1)
+	s.speechRecognizer.Recognized(func(event speech.SpeechRecognitionEventArgs) {
+		defer event.Close()
+
+		if event.Result.Reason == common.NoMatch {
+			slog.Error("no match")
+			return
+		}
+
+		if event.Result.Reason == common.Canceled {
+			slog.Error("canceled")
+			return
+		}
+
+		if len(event.Result.Text) == 0 {
+			slog.Error("empty result")
+			return
+		}
+
+		segmentsCh <- transcribe.Segment{
+			Text:    event.Result.Text,
+			StartTS: int64(event.Result.Offset.Seconds() * 1000),
+			EndTS:   int64(event.Result.Offset.Seconds()*1000 + event.Result.Duration.Seconds()*1000),
+		}
+	})
+
+	err := <-s.speechRecognizer.StartContinuousRecognitionAsync()
+	if err != nil {
+		return nil, fmt.Errorf("failed to start recognizer: %w", err)
+	}
+
+	go func() {
+		defer func() {
+			err := <-s.speechRecognizer.StopContinuousRecognitionAsync()
+			if err != nil {
+				slog.Error("failed to stop recognizer", slog.String("err", err.Error()))
+			}
+			defer close(segmentsCh)
+		}()
+
+		for samples := range samplesCh {
+			if err := s.audioStream.Write(f32PCMToWAV(samples)); err != nil {
+				slog.Error("failed to write audio data", slog.String("err", err.Error()))
+				break
+			}
+		}
+	}()
+
+	return segmentsCh, nil
+}
+
+func (s *SpeechRecognizer) Transcribe(samples []float32) ([]transcribe.Segment, string, error) {
+	// TODO: we should likely re-use the same session throughout a track transcription to optimize
+	// resources a bit.
```

Comment:
That's a good question -- why are you creating a new one here? I see you have stored one, and using that for all the async calls, so you can reuse it, right?

Reply:
@cpoile Truth is I've spent too many hours playing with that TODO, implemented it at least twice and even considered changing it with a `DON' TRY THIS` message.

The library is just a bit bugged. Re-using the client is recommended but I don't think it works properly through their Golang wrapper because everything relies on a stream which can't be flushed so you end up with data loss. And if you close the stream then you need to re-initialize everything like we do, so we are a bit stuck. 

I think the optimal solution may be to extend the interface and pass an audio reader to this method instead of the chunks we create since we are dealing with post-transcript.

Reply:
Ah, interesting. Thanks for the context. (If you remember, it might be useful to add some of that reasoning into the comment in another PR?)